src/unitxt/metrics.py

```diff
@@ -399,31 +655,107 @@ def process(self, stream: Stream, stream_name: Optional[str] = None) -> Generato
 
             instances.append(instance)
 
-        for reduction, fields in self.reduction_map.items():
-            assert (
-                reduction in self.implemented_reductions
-            ), f"Reduction {reduction} is not implemented, use one of {self.implemented_reductions}"
+        return instances, global_score
 
-            if reduction == "mean":
-                from statistics import mean
+    def get_group_scores(
+        self, instances: List[dict], field_names: List[str], group_aggregation_func
+    ):
+        """Return a list of group aggregation function value for group_mean reduction.
 
-                for field_name in fields:
-                    scores = [
-                        instance["score"]["instance"][field_name]
-                        for instance in instances
-                    ]
-                    global_score[field_name] = mean(scores)
-                    if field_name == self.main_score:
-                        global_score["score"] = global_score[field_name]
-                        global_score["score_name"] = self.main_score
+        Args:
+            instances: List of observation instances with instance-level scores (fields) computed.
+            field_names: List of instance score names in each instance to apply the aggregation function.
+            group_aggregation_func: Callable aggregation function accepting a list of numeric scores, or a dict of variant types scores,
+                and returning a single score.
 
-                confidence_interval = self.score_based_confidence_interval(
-                    instances=instances
-                )
-                global_score.update(confidence_interval)
+        Returns:
+            List of dicts, each corresponding to a group of instances (defined by grouping_field),
+                with a group score for each field_name
+        """
+        from collections import defaultdict
+
+        # three-level defaultdict:
+        # first is the grouping, second is the field name, the third is the variant_type (by default 'original')
+        group_to_instance_scores = defaultdict(
```

Comment:
because group_aggregation_function operates on each group scores individually.  It is aggregating the scores in a group, not aggregating the overall scores of multiple groups.  My aim is that people do not have to specify the grouping within this aggregation function.